✨(deploy) manage cronjob object

lunika opened the pull request:
## Purpose

In some applications we would like to add cronjob objects. In order to
manage them we have a new regex mapping cronjob files (cronjob_*.yml.j2)
and then create them in the manage_app task. They are creating before
DeploymentConfig.

## Proposal

- [x] register cronjob templates in `get_objects_for_app` task.
- [x] create/delete cronjob object in `manage_app` task.

lunika committed:
✨(deploy) manage cronjob object

In some applications we would like to add cronjob objects. In order to
manage them we have a new regex mapping cronjob files (cronjob_*.yml.j2)
and then create them in the manage_app task. They are created before
DeploymentConfig.

madmatah reviewed (commented):
Do you plan to handle the `switch` problem (to avoid cronjob from multiple stacks to be active simultaneously) in another PR ? I've found that we can use `spec.suspend: true` in a cronjob definition to prevent new jobs from being run.

lunika commented:
> Do you plan to handle the `switch` problem (to avoid cronjob from multiple stacks to be active simultaneously) in another PR ? I've found that we can use `spec.suspend: true` in a cronjob definition to prevent new jobs from being run.

I didn't start working on the switch part, but with what you found I will work on this PR. Thanks

jmaupetit commented on CHANGELOG.md:
```diff
@@ -20,6 +20,7 @@ Versioning](http://semver.org/spec/v2.0.0.html).
   - `marsha`
   - `edxec`
   - `edxapp`
+- Manage cronjob objects
```
```suggestion
- Handle `CronJob` object type
```

lunika committed:
♻️(task) delete CronJob resource in delete_app task

CronJob resource shloud be deleted like any other resources when the
delete_app task is executed. We faced an issue doing this, `CronJob`
doesn't use the apiVersion `v1` but `v1beta1` and this value was
hardcoded in our code. The apiVersion value is now set in the delete_app
task and used by delete_app_objects_kind task.

lunika committed:
✨(task) suspend next and previous CronJobs

Only the CronJob in the current stack should be active, the other ones
are suspended and activated only when they are switched to the current
stack.

lunika commented:
@madmatah switch and delete are handle now

madmatah reviewed (commented):
Can you please provide an example of cronjob template definition in `hello` app ?

It will be useful as an example and it will allow CircleCI to test your ansible tasks.

jmaupetit commented on tasks/switch_cronjobs.yml:
```diff
@@ -0,0 +1,26 @@
+---
+# In a switch, cronjobs in previous stacks must be suspended.
```
```suggestion
# In a switch, cronjobs from the previous stack must be suspended.
```

jmaupetit commented on tasks/switch_cronjobs.yml:
```diff
@@ -0,0 +1,26 @@
+---
+# In a switch, cronjobs in previous stacks must be suspended.
+# CronJobs in the current stack must be activated
```
```suggestion
# and those from the current stack must be activated.
```

sampaccoud commented on CHANGELOG.md:
```diff
@@ -20,6 +20,7 @@ Versioning](http://semver.org/spec/v2.0.0.html).
   - `marsha`
```
In the commit message : 
s/They are creating before DeploymentConfig/They are created before DeploymentConfig

sampaccoud commented on tasks/delete_app.yml:
```diff
@@ -6,15 +6,16 @@
 - name: Iterate over objects to delete with deployment_stamp {{ targeted_deployment_stamp }} for app {{ app.name }}
```
in commit message:
s/shlould/should
s/wah/was
s/ans use/and used
s/delete_app_objects_kinf/delete_app_objects_kind

jmaupetit replied:
:joy: 

![](https://media.giphy.com/media/ZezDqIMwHiqLm/giphy.gif)

lunika replied:
I didn't laugh...

sampaccoud replied:
@jmaupetit so you're the referee on four legs?

sampaccoud commented on tasks/delete_app_objects_kind.yml:
```diff
@@ -1,30 +1,31 @@
 # Delete a kind of objects for the targeted deployment_stamp
-- name: Lookup {{ kind }} to delete with deployment_stamp {{ targeted_deployment_stamp }} for app {{ app.name }}
+- name: Lookup {{ resource_to_delete.kind }} to delete with deployment_stamp {{ targeted_deployment_stamp }} for app {{ app.name }}
```
Why don't you remove the `version` selector so that we delete all objects of this kind whatever their version?

lunika replied:
because by default the value is `v1` see https://docs.ansible.com/ansible/latest/plugins/lookup/k8s.html

sampaccoud replied:
wow that sucks. Have you tried setting it to null?

lunika replied:
No, do you think it's better than what I made ?

sampaccoud replied:
well it would be simpler and would avoid looping on all versions when you get several?

lunika replied:
I don't think we loop on all versions, we use only the one defined in the hashmap

sampaccoud replied:
Ok up to you but it is also uglier that you have to pass a whole object to define what you want to delete... so if the version could be ignored it would be better imo. If null doesn't work then ok lah.

lunika replied:
I tried and no resources are deleted, the `delete_app` fails with a timeout

sampaccoud commented on apps/hello/templates/services/app/cronjob_hello.yml.j2:
```diff
@@ -0,0 +1,33 @@
+apiVersion: batch/v1beta1
```
in commit message:
s/currect/correct
s/the other one/the other ones
s/active only if the are switc in the current task/activated only when they are switched to the current stack
